### PR TITLE
Document `NodeId`

### PR DESCRIPTION
**Pull request by camelid**

### CODE REVIEW

File: compiler/rustc_ast/src/node_id.rs [hunk 1]
```diff
@@ -2,6 +2,12 @@ use rustc_span::ExpnId;
 use std::fmt;
 
 rustc_index::newtype_index! {
+    /// Identifies an AST node.
+    ///
+    /// This identifies top-level definitions, expressions, and everything in between.
+    /// This is later split into [`DefId`] and `HirId` for the HIR.
```
Comment (Aaron1011) [inline compiler/rustc_ast/src/node_id.rs:8]: Could you say "converted" instead of "split"?

Comment (camelid) [reply compiler/rustc_ast/src/node_id.rs]: "converted" didn't sound quite right to me, so I did "turned into" instead.

File: compiler/rustc_ast/src/node_id.rs [hunk 2]
```diff
@@ -2,19 +2,25 @@ use rustc_span::ExpnId;
...
 rustc_data_structures::define_id_collections!(NodeMap, NodeSet, NodeId);
 
-/// `NodeId` used to represent the root of the crate.
+/// The [`NodeId`] used to represent the root of the crate.
 pub const CRATE_NODE_ID: NodeId = NodeId::from_u32(0);
 
 /// When parsing and doing expansions, we initially give all AST nodes this AST
-/// node value. Then later, during expansion, we renumber them to have small,
-/// positive ids.
+/// [`NodeId`]. Then later, during expansion, we renumber them to have small,
```
Comment (camelid) [inline compiler/rustc_ast/src/node_id.rs:22]: This part is confusing to me: it says that "when parsing and doing expansions" we use a dummy ID, but "then later, during expansion" we create real IDs. Does it mean that at the beginning of expansions we use dummy IDs but a later phase of it we create real ones?

Comment (Aaron1011) [reply compiler/rustc_ast/src/node_id.rs:22]: That's correct - the creation of real `NodeId`s is done here: https://github.com/rust-lang/rust/blob/9bc8b00b4a4e38ccbc3aeec2c123538973c67eba/compiler/rustc_expand/src/expand.rs#L1702-L1707

**Comment by camelid** [comment]
I was planning on clarifying the part about the dummy IDs a bit, and it might make sense for Aaron1011 to take another look too once I do that. @bors r-

**Comment by camelid** [comment]
Ready for a final review!

**Pull request merged**